core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java

```diff
@@ -68,38 +72,46 @@ public static ParseItem get(final ParseGraph graph, final Token definition) {
     public static ImmutableList<Value> getAllValues(final ParseGraph graph, final Token definition) {
         checkNotNull(graph, "graph");
         checkNotNull(definition, "definition");
-        return getAllValuesRecursive(graph, definition);
+        return reverse(getAllValuesRecursive(ImmutableList.create(graph), new ImmutableList<>(), definition).computeResult());
     }
 
-    private static ImmutableList<Value> getAllValuesRecursive(final ParseGraph graph, final Token definition) {
-        if (graph.isEmpty()) { return new ImmutableList<>(); }
-        final ImmutableList<Value> tailResults = getAllValuesRecursive(graph.tail, definition);
+    private static SafeTrampoline<ImmutableList<Value>> getAllValuesRecursive(final ImmutableList<ParseGraph> graphList, final ImmutableList<Value> values, final Token definition) {
+        if (graphList.isEmpty()) { return complete(() -> values); }
+        final ParseGraph graph = graphList.head;
+        if (graph.isEmpty()) { return intermediate(() -> getAllValuesRecursive(graphList.tail, values, definition)); }
         final ParseItem head = graph.head;
         if (head.isValue() && head.asValue().definition.equals(definition)) {
-            return tailResults.add(head.asValue());
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values.add(head.asValue()), definition));
         }
         if (head.isGraph()) {
-            return tailResults.add(getAllValuesRecursive(head.asGraph(), definition));
+            return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail).add(graph.head.asGraph()), values, definition));
         }
-        return tailResults;
+        return intermediate(() -> getAllValuesRecursive(graphList.tail.add(graph.tail), values, definition));
     }
 
     public static ImmutableList<ParseItem> getAllRoots(final ParseGraph graph, final Token definition) {
-        checkNotNull(graph, "graph");
-        checkNotNull(definition, "definition");
-        return getAllRootsRecursive(graph, null, definition);
+        return getAllRootsRecursive(ImmutableList.create(new Pair(checkNotNull(graph, "graph"), null)), checkNotNull(definition, "definition"), new ImmutableList<>()).computeResult();
     }
 
-    private static ImmutableList<ParseItem> getAllRootsRecursive(final ParseItem item, final ParseGraph parent, final Token definition) {
-        final ImmutableList<ParseItem> result = item.getDefinition().equals(definition) && (parent == null || !parent.getDefinition().equals(definition))
-            ? ImmutableList.create(item)
-            : new ImmutableList<>();
+    private static SafeTrampoline<ImmutableList<ParseItem>> getAllRootsRecursive(final ImmutableList<Pair> list, final Token definition, final ImmutableList<ParseItem> result) {
```

Comment:
Renamed `list` to `backlog` (because it's the list of roots that still need to be checked) and `result` to `rootList` (because it's the list of found roots).